File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        int length = 0;
        ListNode temp = head;

        // Step 1: Count the total length of the list
        while (temp != null) {
            length++;
            temp = temp.next;
        }

        int partSize = length / k;
        int extra = length % k;

        ListNode[] result = new ListNode[k];
        temp = head;

        for (int i = 0; i < k && temp != null; i++) {
            result[i] = temp;
            int currentPartSize = partSize + (extra > 0 ? 1 : 0);
            extra--;

            // Move to the end of the current part
            for (int j = 1; j < currentPartSize; j++) {
                temp = temp.next;
            }

            // Disconnect current part from the rest of the list
            ListNode nextPart = temp.next;
            temp.next = null;
            temp = nextPart;
        }

        return result;
    }
}

File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {

           int l =0; 
        ListNode* temp = head; 
        while(temp){
            l++; 
            temp = temp->next;
        }
        int a = l/k; 
        int b = l%k;

        vector<ListNode*>ans(k,NULL); 
        temp = head; 
        ListNode* prev = NULL;

        for( int i = 0; temp && i < k ; i++){
            ans[i] = temp; 

            for( int count = 1 ; count<= a + (b > 0 ? 1:0) ; count++ ){
                prev = temp; 
                temp = temp->next;
            }
            prev->next = NULL; 
            b--;
        }
        return ans; 

        
    }
};